src/Processors/Streaming/ResizeProcessor.cpp

```diff
@@ -0,0 +1,456 @@
+#include <Processors/Streaming/ResizeProcessor.h>
+
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+}
+
+namespace Streaming
+{
+ShrinkResizeProcessor::ShrinkResizeProcessor(const Block & header, size_t num_inputs)
+    : IProcessor(InputPorts(num_inputs, header), OutputPorts(1, header))
+    , log(&Poco::Logger::get("ShrinkResizeProcessor"))
+{
+    assert(num_inputs > 0);
+}
+
+IProcessor::Status ShrinkResizeProcessor::prepare(const PortNumbers & updated_inputs, const PortNumbers & /*updated_outputs*/)
```

Comment:
I need a bit of explanation about watermarks to understand this and `updateAndAlignWatermark` functions.

- Is it expected that the different inputs has the same source of watermarks, so a single `WatermarkStamper` before an `ExpandResizeProcessor`?
- Can data from different watermarks be mixed? E.g. we have a 5s periodic emit query. Can chunks before and after a water be sent to the output without sending the watermark? Isn't the goal of watermarks to keep the data and the outputted result "in sync"?

Reply:
> I need a bit of explanation about watermarks to understand this and `updateAndAlignWatermark` functions.
> 
> * Is it expected that the different inputs has the same source of watermarks, so a single `WatermarkStamper` before an `ExpandResizeProcessor`?

Different inputs have different source of watermarks (timestamps). Watermarks are stamped (observed) independently for each of time in the code. However, when we reach a Resize, Aggregate, Join operators, we will need align them (which fundamentally is trying to align the timestamps - what time it is now). After alignment, we established a timestamp for the the down stream query pipeline.

> * Can data from different watermarks be mixed? E.g. we have a 5s periodic emit query. Can chunks before and after a water be sent to the output without sending the watermark? Isn't the goal of watermarks to keep the data and the outputted result "in sync"?

Watermark can be mixed and will need re-alignment when we hit a Resize etc operator like mentioned above. The sole goal of watermark is answering `what time is it` and all of the aggr, join emit, state GC etc are based on the watermarks.